### ISSUE TITLE
generateMultiplyDivideExplanation can use previous $1 value, resulting in incorrect value

### ISSUE THREAD

**Issue by mole6e23** [WeBWorK-for-Chemistry/WeBWorKChemistry#1]
On line 853 of contextInexactValue.pl inside the subroutine generateMultiplyDivideExplanation, the line

	if (defined $1){
 
 is used to detect whether the previous substitution

	$trimmedData = $originalData =~ s/^\s*-?\s*(0[.,]0*)//r;

was successful and apply leading zeros to the value if so. However, I found cases where the previous regex substitution failed (appropriately) and the value in $1 was still defined from the previous regex (somewhere in the program). For example, if I had the value 820450970 rounded to 7 significant figures, the system would return 008204510 as the answer (see attached screen shot).

I changed the test to

if ($trimmedData ne $originalData ){

to see if a substitution was made and that seemed to fix it for this case, but I didn't widely test this.

There are several other places where "defined $1" is used without testing for a successful regex. In these cases, it is quite possible there are more of these errors since $1 can be defined from a previous regex.

![Screen Shot 2023-06-08 at 12 31 52 PM](https://github.com/WeBWorK-for-Chemistry/Macros/assets/136017750/b1cd0f4e-0427-40e2-9875-948426383164)

**Comment by mole6e23** [comment]
As a side note, it turns out that the answer given when I used the new test seemed correct, but was still actually incorrect as it put the answer as 82045100 which is numerically correct, but doesn't have 7 significant figures (in this case the system should have forced to scientific notation). The incorrect rounding/scientific notation is a different error than the one described in this issue.

**Comment by limefrogyank** [comment]
Thanks for these issues!  Sorry for the late reply, but I'm finally gearing up to work on this full-time again.

**Comment by limefrogyank** [comment]
I couldn't exactly replicate your exact example, but I did see some weird behavior and I believe I came up with a way to fix it.

First, there were a bunch of regular regex matching statements where I was not checking for success.  Those were easy to fix.  

But there was one in the generateSfRoundingExplanation function that was special.  I was doing a non-destructive substitution with the 'r' flag.  I couldn't figure out the way to check for success on that one, so I decided I'd have to first do a plain regex matching check for success.   If successful, then do the substitution using the same pattern.  But that meant that any groups ($1) would be undefined from the first regex pattern (or filled with the group match).  

```
    if ($originalData =~ /^\s*-?\s*(0[.,]0*)/){
        $trimmedData = $originalData =~ s/^\s*-?\s*(0[.,]0*)//r;
        if (defined $1){
            $leadingZeros = $1;
        }
    } else {
        $trimmedData = $originalData;
```
If you see any more problems with this, let me know!  And thank you for diagnosing this!  It would have been awful to track this down myself.  I am still quite naive about Perl in a few places.
